File: demo/android/app/src/main/kotlin/com/mkparser/demo/ui/render/MkBlockParser.kt
package com.mkparser.demo.ui.render

import androidx.compose.ui.graphics.Color
import androidx.compose.ui.text.AnnotatedString
import androidx.compose.ui.text.SpanStyle
import androidx.compose.ui.text.buildAnnotatedString
import androidx.compose.ui.text.font.FontFamily
import androidx.compose.ui.text.font.FontStyle
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextDecoration
import com.mkparser.MkParser
import com.mkparser.NodeType

/**
 * Parses a Markdown string into a flat [List<MkBlock>] by driving [MkParser]
 * and processing its push-events via a state machine.
 *
 * Block IDs are stable across re-parses (type + ordinal), so LazyColumn
 * can diff efficiently and only recompose changed items.
 */
object MkBlockParser {

    fun parse(markdown: String, isStreaming: Boolean = false): List<MkBlock> {
        if (markdown.isBlank()) return emptyList()
        val builder = BlockBuilder()
        val parser = MkParser(
            onNodeOpen  = { type, flags, attr -> builder.onOpen(type, flags, attr) },
            onNodeClose = { type              -> builder.onClose(type) },
            onText      = { text             -> builder.onText(text) },
        )
        try {
            parser.feed(markdown).finish()
        } finally {
            parser.destroy()
        }
        val blocks = builder.build()
        return if (isStreaming) blocks.markLastStreaming() else blocks
    }
}

// ── [F03] Incremental streaming parser ────────────────────────────────────────
//
// Holds a long-lived MkParser + BlockBuilder pair so each new token is fed
// only once instead of re-parsing the entire accumulated document.
// This turns the streaming render pass from O(n²) into O(n).

class MkStreamingParser {
    private val builder = BlockBuilder()
    private var parser: MkParser? = MkParser(
        onNodeOpen  = { type, flags, attr -> builder.onOpen(type, flags, attr) },
        onNodeClose = { type              -> builder.onClose(type) },
        onText      = { text             -> builder.onText(text) },
    )

    /** Feed the next chunk and return the current block list (marked streaming). */
    fun feed(chunk: String): List<MkBlock> {
        parser?.feed(chunk)
        return builder.build().markLastStreaming()
    }

    /** Signal end-of-stream and return the final block list. */
    fun finish(): List<MkBlock> {
        parser?.finish()?.destroy()
        parser = null
        return builder.build()
    }
}

// ── Colours used for inline spans ────────────────────────────────────────────

private val colLink   = Color(0xFF89DCEB)
private val colCode   = Color(0xFFCBA6F7)
private val colCodeBg = Color(0xFF313244)

// ── State machine ─────────────────────────────────────────────────────────────

private class BlockBuilder {

    private val blocks   = mutableListOf<MkBlock>()
    private val typeCounts = mutableMapOf<String, Int>()

    // ── Block-level stack ─────────────────────────────────────────────────────

    private data class Frame(val type: Int, val flags: Int, val attr: String?, val id: String)
    private val stack = ArrayDeque<Frame>()

    // Single shared inline builder (owned by the innermost content block).
    private var ib: AnnotatedString.Builder? = null

    // How many AnnotatedString pushes each open inline span created.
    private val ibPushes = ArrayDeque<Int>()

    // ── List context ──────────────────────────────────────────────────────────

    private data class ListCtx(val ordered: Boolean, var itemNum: Int = 0)
    private val listStack = ArrayDeque<ListCtx>()

    // Task state set by a TASK_LIST_ITEM leaf marker, consumed by LIST_ITEM close.
    private var pendingTaskState: Int = 0

    // ── Table context ─────────────────────────────────────────────────────────

    private val tableHeaders = mutableListOf<AnnotatedString>()
    private val tableRows    = mutableListOf<List<AnnotatedString>>()
    private val currentRow   = mutableListOf<AnnotatedString>()
    private var inTableHead  = false

    // ── Helpers ───────────────────────────────────────────────────────────────

    private fun newId(typeName: String): String {
        val n = typeCounts.getOrDefault(typeName, 0)
        typeCounts[typeName] = n + 1
        return "${typeName}_$n"
    }

    /** Block types that "own" the inline builder — PARAGRAPH inside them
     *  should not create its own builder or emit its own block. */
    private val containerTypes = setOf(
        NodeType.LIST_ITEM, NodeType.TASK_LIST_ITEM,
        NodeType.BLOCK_QUOTE, NodeType.TABLE_CELL,
    )

    private fun insideContainer(): Boolean =
        stack.any { it.type in containerTypes }

    // ── Public event handlers ─────────────────────────────────────────────────

    fun onOpen(type: Int, flags: Int, attr: String?) {
        stack.addLast(Frame(type, flags, attr, newId(typeKey(type))))

        when (type) {
            NodeType.HEADING -> ib = AnnotatedString.Builder()

            NodeType.CODE_BLOCK -> ib = AnnotatedString.Builder()

            NodeType.BLOCK_QUOTE -> ib = AnnotatedString.Builder()

            NodeType.PARAGRAPH -> {
                if (!insideContainer()) ib = AnnotatedString.Builder()
            }

            NodeType.LIST_ITEM -> {
                pendingTaskState = 0        // reset; TASK_LIST_ITEM sibling will set it
                ib = AnnotatedString.Builder()
            }

            NodeType.TASK_LIST_ITEM -> {
                // TASK_LIST_ITEM is a leaf marker node (sibling of PARAGRAPH inside LIST_ITEM).
                // Store its checked state so LIST_ITEM close can use it.
                // Do NOT overwrite ib — LIST_ITEM already owns the builder.
                // If ib is null (standalone TASK_LIST_ITEM without a LIST_ITEM parent),
                // create a builder as fallback so text is not lost.
                pendingTaskState = flags
                if (ib == null) ib = AnnotatedString.Builder()
            }

            NodeType.LIST -> listStack.addLast(ListCtx(ordered = flags == 1))

            NodeType.TABLE -> {
                tableHeaders.clear()
                tableRows.clear()
            }

            NodeType.TABLE_HEAD -> inTableHead = true

            NodeType.TABLE_ROW -> if (!inTableHead) currentRow.clear()

            NodeType.TABLE_CELL -> ib = AnnotatedString.Builder()

            NodeType.THEMATIC_BREAK -> { /* leaf, handled on close */ }

            // ── Inline spans ──────────────────────────────────────────────────

            NodeType.STRONG -> ibPushes.addLast(
                ib?.let { it.pushStyle(SpanStyle(fontWeight = FontWeight.Bold)); 1 } ?: 0
            )
            NodeType.EMPHASIS -> ibPushes.addLast(
                ib?.let { it.pushStyle(SpanStyle(fontStyle = FontStyle.Italic)); 1 } ?: 0
            )
            NodeType.STRIKETHROUGH -> ibPushes.addLast(
                ib?.let { it.pushStyle(SpanStyle(textDecoration = TextDecoration.LineThrough)); 1 } ?: 0
            )
            NodeType.INLINE_CODE -> ibPushes.addLast(
                ib?.let {
                    it.pushStyle(SpanStyle(fontFamily = FontFamily.Monospace,
                        color = colCode, background = colCodeBg))
                    1
                } ?: 0
            )
            NodeType.LINK,
            NodeType.AUTO_LINK -> {
                val url = attr ?: ""
                val pushed = ib?.let { b ->
                    b.pushStyle(SpanStyle(color = colLink, textDecoration = TextDecoration.Underline))
                    if (url.isNotEmpty()) { b.pushStringAnnotation("URL", url); 2 } else 1
                } ?: 0
                ibPushes.addLast(pushed)
            }

            NodeType.SOFT_BREAK -> ib?.append(" ")
            NodeType.HARD_BREAK -> ib?.append("\n")
        }
    }

    fun onClose(type: Int) {
        val frame = stack.removeLastOrNull() ?: return

        when (type) {

            // ── Inline span pop ───────────────────────────────────────────────

            NodeType.STRONG,
            NodeType.EMPHASIS,
            NodeType.STRIKETHROUGH,
            NodeType.INLINE_CODE,
            NodeType.LINK,
            NodeType.AUTO_LINK -> {
                val count = ibPushes.removeLastOrNull() ?: 0
                repeat(count) { ib?.pop() }
            }

            // ── Block emit ────────────────────────────────────────────────────

            NodeType.HEADING -> {
                val c = ib?.toAnnotatedString() ?: AnnotatedString(""); ib = null
                blocks += MkBlock.Heading(frame.id, frame.flags, c)
            }

            NodeType.PARAGRAPH -> {
                if (!insideContainer()) {
                    val c = ib?.toAnnotatedString() ?: AnnotatedString(""); ib = null
                    blocks += MkBlock.Paragraph(frame.id, c)
                }
                // else: container still owns the builder
            }

            NodeType.CODE_BLOCK -> {
                val code = ib?.toAnnotatedString()?.text?.trimEnd('\n') ?: ""; ib = null
                blocks += MkBlock.FencedCode(frame.id, frame.attr ?: "", code)
            }

            NodeType.BLOCK_QUOTE -> {
                val c = ib?.toAnnotatedString() ?: AnnotatedString(""); ib = null
                blocks += MkBlock.BlockQuote(frame.id, c)
            }

            NodeType.LIST_ITEM -> {
                val c = ib?.toAnnotatedString() ?: AnnotatedString(""); ib = null
                val ctx = listStack.lastOrNull()
                if (ctx?.ordered == true) {
                    ctx.itemNum++
                    blocks += MkBlock.OrderedItem(frame.id, listStack.size - 1, ctx.itemNum, c)
                } else {
                    // pendingTaskState is set by a TASK_LIST_ITEM sibling marker (1=unchecked,2=checked).
                    // frame.flags carries mk_node_list_item_task_state as fallback.
                    val taskState = if (pendingTaskState != 0) pendingTaskState else frame.flags
                    pendingTaskState = 0
                    blocks += MkBlock.BulletItem(frame.id, listStack.size - 1, c, taskState = taskState)
                }
            }

            NodeType.TASK_LIST_ITEM -> {
                // If there is a LIST_ITEM ancestor, this is a leaf marker — it emits nothing.
                // The LIST_ITEM close handler will emit the block using pendingTaskState.
                // If standalone (no LIST_ITEM parent), emit as a regular bullet block.
                val hasListItemParent = stack.any { it.type == NodeType.LIST_ITEM }
                if (!hasListItemParent) {
                    val c = ib?.toAnnotatedString() ?: AnnotatedString(""); ib = null
                    blocks += MkBlock.BulletItem(frame.id, listStack.size - 1, c, taskState = frame.flags)
                }
                // else: do nothing — LIST_ITEM close handles emission
            }

            NodeType.LIST -> listStack.removeLastOrNull()

            NodeType.THEMATIC_BREAK -> blocks += MkBlock.ThematicBreak(frame.id)

            NodeType.TABLE_CELL -> {
                val c = ib?.toAnnotatedString() ?: AnnotatedString(""); ib = null
                if (inTableHead) tableHeaders += c else currentRow += c
            }

            NodeType.TABLE_ROW -> {
                if (!inTableHead) { tableRows += currentRow.toList(); currentRow.clear() }
            }

            NodeType.TABLE_HEAD -> inTableHead = false

            NodeType.TABLE -> blocks += MkBlock.TableBlock(
                frame.id, tableHeaders.toList(), tableRows.toList()
            )
        }
    }

    fun onText(text: String) { ib?.append(text) }

    fun build(): List<MkBlock> = blocks.toList()

    // Stable key prefix per block type
    private fun typeKey(type: Int) = when (type) {
        NodeType.HEADING        -> "H"
        NodeType.PARAGRAPH      -> "P"
        NodeType.CODE_BLOCK     -> "CODE"
        NodeType.BLOCK_QUOTE    -> "BQ"
        NodeType.LIST_ITEM,
        NodeType.TASK_LIST_ITEM -> "LI"
        NodeType.THEMATIC_BREAK -> "HR"
        NodeType.TABLE          -> "TBL"
        NodeType.TABLE_CELL     -> "TD"
        else                    -> "X"
    }
}

File: demo/ios/App/render/MkBlockParser.swift
// MkBlockParser.swift — Push-event state machine: MkParser → [MkBlock]
//
// Logic mirrors the Android BlockBuilder in MkBlockParser.kt exactly.
// Converts mk_parser push-events into a flat list of MkBlock values
// for direct rendering by MkRenderView.

import Foundation
import SwiftUI
import MkParser   // bindings/ios/MkParser.swift (via Package.swift)

// ── Colours for inline spans (Catppuccin Mocha palette) ─────────────────────

private let colLink   = Color(red: 0x89/255.0, green: 0xDC/255.0, blue: 0xEB/255.0)
private let colCode   = Color(red: 0xCB/255.0, green: 0xA6/255.0, blue: 0xF7/255.0)
private let colCodeBg = Color(red: 0x31/255.0, green: 0x32/255.0, blue: 0x44/255.0)

// ── Public parser ─────────────────────────────────────────────────────────────

enum MkBlockParser {

    /// Parse a complete markdown string and return a flat block list.
    static func parse(_ markdown: String, isStreaming: Bool = false) -> [MkBlock] {
        guard !markdown.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty else { return [] }
        let builder = BlockBuilder()
        let parser  = try? MkParser(
            onNodeOpen:  { type, info  in builder.onOpen(type: type, info: info)  },
            onNodeClose: { type, _     in builder.onClose(type: type) },
            onText:      { text        in builder.onText(text) },
            onModify:    { _, _        in }
        )
        try? parser?.feed(markdown)
        try? parser?.finish()
        parser?.destroy()
        let blocks = builder.build()
        return isStreaming ? blocks.markLastStreaming() : blocks
    }
}

// ── Internal state machine ────────────────────────────────────────────────────

private class BlockBuilder {

    private var blocks: [MkBlock] = []
    private var typeCounts: [String: Int] = [:]

    // Block-level stack
    private struct Frame {
        let type: MkNodeType
        let info: MkNodeInfo
        let id:   String
    }
    private var stack: [Frame] = []

    // Inline text builder (owned by the innermost content frame)
    private var ib: AttributedString?

    // Stack of inline span closures (to "pop" applied styles on close)
    private var ibRangeStarts: [(AttributedString.Index, Int)] = []
                              // (start index, push-depth) for pop

    // List context
    private struct ListCtx {
        let ordered: Bool
        var itemNum: Int = 0
    }
    private var listStack: [ListCtx] = []

    // Task state from TASK_LIST_ITEM sibling, consumed by LIST_ITEM close
    private var pendingTaskState: Int = 0

    // Table state
    private var tableHeaders:  [AttributedString] = []
    private var tableRows:     [[AttributedString]] = []
    private var currentRow:    [AttributedString] = []
    private var inTableHead    = false

    // ── ID generation ─────────────────────────────────────────────────────────

    private func newId(_ prefix: String) -> String {
        let n = typeCounts[prefix, default: 0]
        typeCounts[prefix] = n + 1
        return "\(prefix)_\(n)"
    }

    // ── Container check ───────────────────────────────────────────────────────

    private let containerTypes: Set<MkNodeType> = [
        .listItem, .taskListItem, .blockQuote, .tableCell
    ]
    private func insideContainer() -> Bool {
        stack.contains { containerTypes.contains($0.type) }
    }

    // ── Event handlers ────────────────────────────────────────────────────────

    func onOpen(type: MkNodeType, info: MkNodeInfo) {
        let id = newId(typeKey(type))
        stack.append(Frame(type: type, info: info, id: id))

        switch type {

        // Block openers that own the inline builder
        case .heading:
            ib = AttributedString()

        case .codeBlock:
            ib = AttributedString()

        case .blockQuote:
            ib = AttributedString()

        case .paragraph:
            if !insideContainer() { ib = AttributedString() }

        case .listItem:
            pendingTaskState = 0
            ib = AttributedString()

        case .taskListItem:
            // Leaf marker — store task state; LIST_ITEM owns the builder
            pendingTaskState = info.taskState == .checked ? 2 :
                               info.taskState == .unchecked ? 1 : 0
            if ib == nil { ib = AttributedString() }

        case .list:
            listStack.append(ListCtx(ordered: info.ordered ?? false))

        case .table:
            tableHeaders = []; tableRows = []

        case .tableHead:
            inTableHead = true

        case .tableRow:
            if !inTableHead { currentRow = [] }

        case .tableCell:
            ib = AttributedString()

        case .thematicBreak:
            break  // leaf — handled on close

        // ── Inline span opens ──────────────────────────────────────────────────

        case .strong:
            guard var current = ib else { ibRangeStarts.append((AttributedString.Index(utf16Offset: 0, in: AttributedString()), 0)); break }
            let start = current.endIndex
            ib = current
            ibRangeStarts.append((start, 1))

        case .emphasis:
            guard var current = ib else { ibRangeStarts.append((AttributedString.Index(utf16Offset: 0, in: AttributedString()), 0)); break }
            let start = current.endIndex
            ib = current
            ibRangeStarts.append((start, 1))

        case .strikethrough:
            guard var current = ib else { ibRangeStarts.append((AttributedString.Index(utf16Offset: 0, in: AttributedString()), 0)); break }
            let start = current.endIndex
            ib = current
            ibRangeStarts.append((start, 1))

        case .inlineCode:
            guard var current = ib else { ibRangeStarts.append((AttributedString.Index(utf16Offset: 0, in: AttributedString()), 0)); break }
            let start = current.endIndex
            ib = current
            ibRangeStarts.append((start, 1))

        case .link, .autoLink:
            guard var current = ib else { ibRangeStarts.append((AttributedString.Index(utf16Offset: 0, in: AttributedString()), 0)); break }
            let start = current.endIndex
            ib = current
            ibRangeStarts.append((start, 1))

        case .softBreak:
            ib?.append(AttributedString(" "))

        case .hardBreak:
            ib?.append(AttributedString("\n"))

        default:
            break
        }
    }

    func onClose(type: MkNodeType) {
        guard let frame = stack.last else { return }
        stack.removeLast()

        switch type {

        // ── Inline span close — apply accumulated style to the range ──────────

        case .strong:
            applyInlineStyle(frame: frame, type: .strong)

        case .emphasis:
            applyInlineStyle(frame: frame, type: .emphasis)

        case .strikethrough:
            applyInlineStyle(frame: frame, type: .strikethrough)

        case .inlineCode:
            applyInlineStyle(frame: frame, type: .inlineCode)

        case .link, .autoLink:
            applyInlineStyle(frame: frame, type: .link, href: frame.info.href ?? frame.info.url ?? "")

        // ── Block emitters ────────────────────────────────────────────────────

        case .heading:
            let content = ib ?? AttributedString(); ib = nil
            blocks.append(.heading(id: frame.id, level: Int(frame.info.level ?? 1), content: content))

        case .paragraph:
            if !insideContainer() {
                let content = ib ?? AttributedString(); ib = nil
                blocks.append(.paragraph(id: frame.id, content: content))
            }

        case .codeBlock:
            var text = ib.map { String($0.characters) } ?? ""; ib = nil
            if text.hasSuffix("\n") { text = String(text.dropLast()) }
            blocks.append(.fencedCode(id: frame.id, language: frame.info.lang ?? "", code: text))

        case .blockQuote:
            let content = ib ?? AttributedString(); ib = nil
            blocks.append(.blockQuote(id: frame.id, content: content))

        case .listItem:
            let content = ib ?? AttributedString(); ib = nil
            if let ctx = listStack.last, ctx.ordered {
                listStack[listStack.count - 1].itemNum += 1
                let num = listStack[listStack.count - 1].itemNum
                blocks.append(.orderedItem(id: frame.id, depth: listStack.count - 1, number: num, content: content))
            } else {
                let ts = pendingTaskState != 0 ? pendingTaskState : (frame.info.taskState == .checked ? 2 : frame.info.taskState == .unchecked ? 1 : 0)
                pendingTaskState = 0
                blocks.append(.bulletItem(id: frame.id, depth: listStack.count - 1, taskState: ts, content: content))
            }

        case .taskListItem:
            let hasListItemParent = stack.contains { $0.type == .listItem }
            if !hasListItemParent {
                let content = ib ?? AttributedString(); ib = nil
                let ts = pendingTaskState != 0 ? pendingTaskState : (frame.info.taskState == .checked ? 2 : 1)
                blocks.append(.bulletItem(id: frame.id, depth: listStack.count - 1, taskState: ts, content: content))
            }

        case .list:
            listStack.removeLast()

        case .thematicBreak:
            blocks.append(.thematicBreak(id: frame.id))

        case .tableCell:
            let content = ib ?? AttributedString(); ib = nil
            if inTableHead { tableHeaders.append(content) } else { currentRow.append(content) }

        case .tableRow:
            if !inTableHead { tableRows.append(currentRow); currentRow = [] }

        case .tableHead:
            inTableHead = false

        case .table:
            blocks.append(.tableBlock(id: frame.id, headers: tableHeaders, rows: tableRows))

        default:
            break
        }
    }

    func onText(_ text: String) {
        ib?.append(AttributedString(text))
    }

    func build() -> [MkBlock] { blocks }

    // ── Style application ─────────────────────────────────────────────────────

    private func applyInlineStyle(frame: Frame, type: MkNodeType, href: String = "") {
        guard var current = ib else { _ = ibRangeStarts.popLast(); return }
        guard let (startIndex, _) = ibRangeStarts.popLast() else { return }

        // Clamp start to valid range
        let safeStart = startIndex <= current.endIndex ? startIndex : current.startIndex
        let range = safeStart..<current.endIndex

        switch type {
        case .strong:
            current[range].font = .body.bold()
        case .emphasis:
            current[range].font = .body.italic()
        case .strikethrough:
            current[range].strikethroughStyle = Text.LineStyle(pattern: .solid)
        case .inlineCode:
            current[range].font = .system(.body, design: .monospaced)
            current[range].foregroundColor = UIColor(colCode)
            current[range].backgroundColor = UIColor(colCodeBg)
        case .link, .autoLink:
            current[range].foregroundColor = UIColor(colLink)
            current[range].underlineStyle  = .single
            if !href.isEmpty, let url = URL(string: href) {
                current[range].link = url
            }
        default:
            break
        }

        ib = current
    }

    // ── Type key ──────────────────────────────────────────────────────────────

    private func typeKey(_ type: MkNodeType) -> String {
        switch type {
        case .heading:      return "H"
        case .paragraph:    return "P"
        case .codeBlock:    return "CODE"
        case .blockQuote:   return "BQ"
        case .listItem, .taskListItem: return "LI"
        case .thematicBreak: return "HR"
        case .table:        return "TBL"
        case .tableCell:    return "TD"
        default:            return "X"
        }
    }
}
